### PR TITLE
Avoid spurious downloads during dependency management

### PR DESCRIPTION
**Pull request by stevendanna**
Before, a URL based source might be downloaded multiple times during the
dependency fetching and lockfile creation. This commit tries to avoid
this by:

1) Memoizing data about the archive to avoid re-fetching the archive

2) Adding a CachedFetcher wrapper around the fetcher class to help
ensure that callers always consult the cache before fetching.

Signed-off-by: Steven Danna steve@chef.io

### CODE REVIEW

File: lib/inspec/dependencies/requirement.rb [hunk 1]
```diff
@@ -94,17 +92,18 @@ def dependencies
...
+      return @profile if ! @profile.nil?
+
       opts = @opts.dup
-      opts[:cache] = @cache
       opts[:backend] = @backend
       if !@dependencies.nil?
         opts[:dependencies] = Inspec::DependencySet.from_array(@dependencies, @cwd, @cache, @backend)
       end
-      @profile ||= Inspec::Profile.for_target(opts, opts)
+      @profile = Inspec::Profile.for_fetcher(fetcher, opts)
```
Comment (chris-rock) [inline lib/inspec/dependencies/requirement.rb:106]: I was thinking about this for a while, since it now deviated from our CLI code. We are using a separate mechanism for `inspec json` now.

Comment (stevendanna) [reply lib/inspec/dependencies/requirement.rb:106]: Can you rephrase this comment? I'm don't understand what you are trying to indicate.

Comment (stevendanna) [reply lib/inspec/dependencies/requirement.rb:106]: Discussion: We are going to leave for_fetcher() but change it so that for_target calls for_fetcher so that it is hard to introduce drifting behavior between the two, this will require the fetcher to expose the original target via an attr_accessor.

Comment (stevendanna) [reply lib/inspec/dependencies/requirement.rb:106]: This is done now.

File: lib/inspec/profile.rb [hunk 2]
```diff
@@ -72,9 +35,14 @@ def self.for_path(path, opts)
       new(reader, opts)
     end
 
+    def self.for_fetcher(fetcher, opts)
+      path, writable = fetcher.fetch
+      for_path(path, opts.merge(target: fetcher.target, writable: writable))
```
Comment (chris-rock) [inline lib/inspec/profile.rb:40]: wouldn't it be better to attach the fetcher to the profile and just read the target then from the fetcher?

Comment (stevendanna) [reply lib/inspec/profile.rb:40]: @chris-rock My plan was to do soemthing like that in a follow up PR since I think that is ultimately what we need to do to remove the Requirement class.

Comment (chris-rock) [reply lib/inspec/profile.rb:40]: Alright, lets go for that approach

**Comment by chris-rock** [comment]
Thanks @stevendanna for this improvement.
